openedx_authz/api/data.py

```diff
@@ -0,0 +1,393 @@
+        while the external key is usually used when initializing from user input or API calls. For example,
+        when creating a role assignment for a content library, the API call would provide the library ID
+        (external_key) and the system would need to determine the correct scope subclass based on the
+        format of the library ID. While when retrieving role assignments from the policy store, the
+        namespaced_key would be used to determine the subclass.
+        """
+        if cls is not ScopeData:
+            return super().__call__(*args, **kwargs)
+
+        if "namespaced_key" in kwargs:
+            scope_cls = cls.get_subclass_by_namespaced_key(kwargs["namespaced_key"])
+            return super(ScopeMeta, scope_cls).__call__(*args, **kwargs)
+
+        if "external_key" in kwargs:
+            scope_cls = cls.get_subclass_by_external_key(kwargs["external_key"])
+            return super(ScopeMeta, scope_cls).__call__(*args, **kwargs)
+
+        return super().__call__(*args, **kwargs)
+
+    @classmethod
+    def get_subclass_by_namespaced_key(mcs, namespaced_key: str) -> Type["ScopeData"]:
+        """Get the appropriate subclass based on the namespace in namespaced_key.
+
+        Args:
+            namespaced_key: The namespaced key (e.g., 'lib^any-library').
+
+        Returns:
+            The subclass of ScopeData corresponding to the namespace, or ScopeData if not found.
+        """
+        # TODO: Default separator, can't access directly from class so made it a constant
+        namespace = namespaced_key.split(AUTHZ_POLICY_ATTRIBUTES_SEPARATOR, 1)[0]
+        return mcs.scope_registry.get(namespace, ScopeData)
+
+    @classmethod
+    def get_subclass_by_external_key(mcs, external_key: str) -> Type["ScopeData"]:
+        """Get the appropriate subclass based on the format of external_key.
+
+        Args:
+            external_key: The external key (e.g., 'lib:any-library').
+
+        Returns:
+            The subclass of ScopeData corresponding to the namespace, or ScopeData if not found.
+        """
+        # Here we need to assume a couple of things:
+        # 1. The external_key is always in the format 'namespace...:other things'. E.g., 'lib:any-library',
+        # even 'course-v1:edX+DemoX+2021_T1'. This won't work for org scopes because they don't explicitly indicate
+        # the namespace in the external key. TODO: We need to handle org scopes differently.
+        # 2. The namespace is always the part before the first separator.
+        # 3. If the namespace is not recognized, we return the base ScopeData class
+        # 4. The subclass implements a validation method to validate the entire key
+        namespace = external_key.split(":", 1)[0]
+        scope_subclass = mcs.scope_registry.get(namespace)
+        if not scope_subclass or not scope_subclass.validate_external_key(external_key):
+            return ScopeData  # Fallback to base class if not found or invalid
```

Comment:
Why return this instead of error? Is there a case where the base class can be useful?